[onert] Make `--shape_run h5` as default option of nnpackage_run when '--load' is provided

hyunsik-yoon opened the pull request:
This make `nnpackage_run` work the following case:
- `--load` is privided
- _no_ `--shape_run` or `--shape_prepare` is provided
- when input shape in h5 and input shape in model are _different_

In such case, 
**shapes in h5 file** is used to set the shape before `nnfw_run()`, which is same with `--shape_run h5` option.

When the input shape in h5 and input shape in model is **same**,
`nnfw_set_input_tensorinfo()` returns without doing anything: https://github.com/Samsung/ONE/blob/master/runtime/onert/api/src/nnfw_api_internal.cc#L492

related to https://github.com/Samsung/ONE/pull/3965#issuecomment-679777002

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

hyunsik-yoon commented:
Test log:

<details>

- `neg` nnpackage has a tensor of shape [4,4]

```

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 --shape_prepare "[0, [4,4]]" --shape_run "[0, [1,1]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg
[Execution] Model input shape will be changed at the start of execute()(index: 0)   <-------- log at https://github.com/Samsung/ONE/pull/4028/files#diff-85791d89f834a5f745e652bd9134fcf3R46

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg
[Execution] Model input shape will be changed at the start of execute()(index: 0)

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 --shape_run h5 | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg
[Execution] Model input shape will be changed at the start of execute()(index: 0)

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 --shape_run "[0, [1,1]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg
[Execution] Model input shape will be changed at the start of execute()(index: 0)

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/4x4/input.h5 | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 --shape_prepare h5 | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/1x1/input.h5 --shape_prepare "[0, [1,1]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/4x4/input.h5 --shape_prepare "[0, [1,1]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/4x4/input.h5 --shape_prepare "[0, [4,4]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/4x4/input.h5 --shape_run "[0, [4,4]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

odroid@odroid:~/work/1/ONE$ ./Product/out/bin/nnpackage_run --nnpackage Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg -l tmp/4x4/input.h5 --shape_prepare "[0, [4,4]]" --shape_run "[0, [4,4]]" | grep "Model input shape"
Package Filename Product/out/unittest_standalone/nnfw_api_gtest_models/neg/neg

```

</details>

glistening commented:
@hyunsik-yoon Running nnpackage_run without any shape setting was the basic use case. Is it broken? Is this PR trying to restore the usage?

glistening commented on tests/tools/nnpackage_run/src/args.cc:
```diff
@@ -216,13 +216,18 @@ void Args::Initialize(void)
          "'h5': read shape(s) from H5 input file. '--load' should also be provided.\n"
-         "'[0, [1, 2], 2, []]': set 0th tensor to [1, 2] and 2nd tensor to [].")
+         "if '--load' option is provided but '--shape_prepare' or '--shape_run' is not provided,\n"
+         "'--shape_run h5' will be used by default.\n"
```
@hyunsik-yoon 

1. Do you mean that input shape from `h5` will be applied to `run()` by default? I expected `prepare()` would be default. 

2. Anyway, too complicated... I am not sure they necessarily supported in nnpackage_run?
I would like to know the concrete use scenario and assumption before further change (including this PR).

3. What does it mean "changing the execution input shape"? What input shape does it use before compilation (=prepare)? Is it safe to resize the execution input size arbitrarily?

hyunsik-yoon commented:
@glistening ,

> @hyunsik-yoon Running nnpackage_run without any shape setting was the basic use case. Is it broken? Is this PR trying to restore the usage?

No. I fixed the title and added more contents into main comment.

hyunsik-yoon commented:
@glistening 

> 1. Do you mean that input shape from `h5` will be applied to `run()` by default? I expected `prepare()` would be default. From that, I guess `run` is more common scenario.

Well, I `run()` by default because changing input shape for `run()` is more common. I guess the following:
- most common: just use input shape in a model (model input shape == input shape in h5)
- less common: prepare -> change input shape -> run -> change input shape -> run ... 
  - This seems be to common in language models.
- least common: change input shape -> prepare -> run
  - When input shape is always same.

> 2. Anyway, too complicated... I am not sure they necessarily supported or could be covered in `nnpackage_run`. I would like to know the concrete use scenario and assumption before further change (including this PR).

- the purpose of `nnpackage_run` is to validate onert. 
- I guess both `--shape_prepare` and `--shape_run` are necessary to check dynamic tensor. Without these options, there is no way to run a model with _unknown dim_. 
- With these two options, explicit shape description (e.g., `--shape_run=[0, [1, 2]]`) and using shape in h5 will make 2 more cases, so 2*2 combination. Plus by providing defalut, there are 2 * 2 + 1 cases.
- I also think these are complex but not sure how to reduce these.

> 3. What does it mean "changing the execution input shape"? What input shape does it use before compilation (=prepare)? 

I will revise the description in more details.

> Does model has no unspecified input shape? 

No if `--shape_run` or `--shape_prepare` is provided properly.

> Is it safe to resize the execution input size arbitrarily?

Not always (e.g., in case of a model `add(input1, input2)`, one input shape could be broadcased into another. If shape is not properly given, operation validator or kernel should throw exception.

glistening commented:
@hyunsik-yoon Thank you for offline explanation and answer here.

1.

Here are my understanding:

There are two cases:

Case 1. Specifying input shape before compilation (=before prepare)
  - Mainly it is required because input has unspecified dim (`-1`).
  - Recent tflite converter does not have `-1` (unspecified) in `shape`. Instead, `tflcvt` sets `1` for unspecified dim and uses `shape_signatures` to denote it is modifiable dim.
  - Currently `onert` ignore `shape_signature`.
  - `nnpackage_run` for random input or h5 input can run without trouble.

Case 2. Adjusting input shape after compilation (=run)
 - Your language model is this case.
 - You want to make Case 2 as default
 - I thought adjusting input shape before compilation may cause problem, however, it seems to be checked and run.


2. 

> - less common: prepare -> change input shape -> run -> change input shape -> run ... 
>  - This seems be to common in language models.

As I told yesterday offline, `nnpackage_run` load `h5` only 1 time.

> the purpose of `nnpackage_run` is to validate onert.

If you would like to cover this test scenario, `nnpackage_run` must provide some way to load a sequence of h5 inputs.
Maybe --nnpackage=input1.h5,input2.h5,input3.h5, ... could be used if you want.

hyunsik-yoon commented:
@glistening, 
For 1, Yes. What you wrote is all correct. We could change `default` to `--shape_prepare-h5`. Actually we don't have many expereice to decide the default yet. However, some models I mentioned offline uses `--shape_run=h5` so I selected this as the default. If setting default value is too early, one other option could be removing the default option.
Please, let me know your opinion.
For 2, I think `prepare -> change input shape -> run` (_only one_ run) is also good way of validating our model and that's how I do nowadays to check if a model with dynamic input tensor runs OK.
What you suggested (multiple h5 files) is a good suggestion, too. I will think about your suggestion not in this PR but in other PR.

wateret commented:
(Sorry I didn't read the thread)

What happens when a user gives `--shape_run h5 --shape_prepare h5`?

hyunsik-yoon commented:
@wateret 
> What happens when a user gives `--shape_run h5` `--shape_prepare h5`?

when `--shape_run h5` or `--shape_prepare h5` is provided with `--load input_file.h5`,
shape from `input_file.h5` is used to call `nnfw_set_input_tensorinfo()`

wateret commented:
@hyunsik-yoon I mean, when both are given at the same time.

hyunsik-yoon commented:
@wateret 
in case of  `--shape_prepare "[0, [2, 2]" --shape_run "[0, [1, 1]"`, shape is set to `[2, 2]` at compile time and then `[1, 1]` at run time. 
However, I found that `--shape_run` disabled `--shape_prepare` when we call with `--shape_prepare h5 --shape_run h5`.

Thinking upon your question, I think ability to handle a situation when `--shape_run` and `--shape_prepare` are given by user at the same time is too much for nnpackage_run.

I will fix and make nnpackage_run print error when both `--shape_run` and `--shape_prepare` are given.

wateret reviewed (approved):
LGTM
